feat: Added GitHub app authentication

galbirk opened the pull request:
In this PR I added the feature of authenticating to GitHub API with github app and allowing to fetch metrics about it.

Tests:
```
$ make test                                                                                                                     
?       github.com/infinityworks/github-exporter        [no test files]
?       github.com/infinityworks/github-exporter/config [no test files]
?       github.com/infinityworks/github-exporter/exporter       [no test files]
?       github.com/infinityworks/github-exporter/http   [no test files]
=== RUN   TestHomepage
Created sequence diagram (2641264843_2166136261.html): /Users/galbirkman/projects/github-exporter/test/.sequence/2641264843_2166136261.html
--- PASS: TestHomepage (0.01s)
=== RUN   TestGithubExporter
time="2023-05-21T16:30:35+03:00" level=info msg="Fetching https://api.github.com/repos/myOrg/myRepo?per_page=100 \n"
time="2023-05-21T16:30:35+03:00" level=info msg="Fetching https://api.github.com/repos/myOrg/myRepo/releases \n"
time="2023-05-21T16:30:35+03:00" level=info msg="Fetching https://api.github.com/repos/myOrg/myRepo/pulls \n"
time="2023-05-21T16:30:35+03:00" level=info msg="API data fetched for repository: https://api.github.com/repos/myOrg/myRepo?per_page=100"
time="2023-05-21T16:30:35+03:00" level=info msg="All Metrics successfully collected"
Created sequence diagram (31181356_2166136261.html): /Users/galbirkman/projects/github-exporter/test/.sequence/31181356_2166136261.html
--- PASS: TestGithubExporter (0.07s)
PASS
ok      github.com/infinityworks/github-exporter/test   1.388s
```

Closes #95

ben-z commented:
@henrymcconville can we merge this PR? Would love to have this feature.

henrymcconville commented on exporter/prometheus.go:
```diff
@@ -46,3 +66,55 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	defer resp.Body.Close()
+	// Triggers if rate-limiting isn't enabled on private Github Enterprise installations
+	if resp.StatusCode == 404 {
+		return false, errors.New("404 Error")
```
IMPORTANT: i'm not too familiar with GH Enterprise but if rate-limiting isn't enabled, then shouldn't this be returning `false, nil`? otherwise it will return from the calling function before it has collected metrics

galbirk replied:
Correct, Done.

henrymcconville commented on exporter/prometheus.go:
```diff
@@ -46,3 +66,55 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	gitHubAppKeyPath := os.Getenv("GITHUB_APP_KEY_PATH")
+	gitHubAppId, _ := strconv.ParseInt(os.Getenv("GITHUB_APP_ID"), 10, 64)
+	gitHubAppInstalaltionId, _ := strconv.ParseInt(os.Getenv("GITHUB_APP_INSTALLATION_ID"), 10, 64)
+	itr, err := ghinstallation.NewKeyFromFile(http.DefaultTransport, gitHubAppId, gitHubAppInstalaltionId, gitHubAppKeyPath)
```
typo:
```suggestion
	itr, err := ghinstallation.NewKeyFromFile(http.DefaultTransport, gitHubAppId, gitHubAppInstallationId, gitHubAppKeyPath)
```

galbirk replied:
Done

henrymcconville commented on exporter/prometheus.go:
```diff
@@ -46,3 +66,55 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+}
+
+func (e *Exporter) reAuth() error {
+	gitHubAppKeyPath := os.Getenv("GITHUB_APP_KEY_PATH")
+	gitHubAppId, _ := strconv.ParseInt(os.Getenv("GITHUB_APP_ID"), 10, 64)
+	gitHubAppInstalaltionId, _ := strconv.ParseInt(os.Getenv("GITHUB_APP_INSTALLATION_ID"), 10, 64)
+	itr, err := ghinstallation.NewKeyFromFile(http.DefaultTransport, gitHubAppId, gitHubAppInstalaltionId, gitHubAppKeyPath)
```
same as above: could we call config.SetAPITokenFromGitHubApp here instead of duplicating code?

henrymcconville commented on exporter/prometheus.go:
```diff
@@ -46,3 +66,55 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+		return false, err
+	}
+
+	defaultLimit := os.Getenv("GITHUB_RATE_LIMIT")
+	if len(defaultLimit) == 0 {
+		defaultLimit = "15000"
+	}
+	defaultLimitInt, err := strconv.ParseInt(defaultLimit, 10, 64)
+	if err != nil {
+		return false, err
+	}
```
could this logic be moved to config.go so we only need to process these vars once on startup?

henrymcconville commented on config/config.go:
```diff
@@ -22,6 +26,7 @@ type Config struct {
 	users         []string
 	apiToken      string
 	targetURLs    []string
+	gitHubApp     bool
```
this looks to be defined but never used. would be good to set this during init then reuse this instead of having to call `os.Getenv` for each use

henrymcconville commented on exporter/prometheus.go:
```diff
@@ -46,3 +66,55 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+		return false, err
+	}
+
+	defaultLimit := os.Getenv("GITHUB_RATE_LIMIT")
+	if len(defaultLimit) == 0 {
+		defaultLimit = "15000"
+	}
```
there's a helper function in the `go-common` library we already use here that can do this for you:
```suggestion
	defaultLimit := cfg.GetEnv("GITHUB_RATE_LIMIT", "15000")
```

henrymcconville reviewed (changes requested):
Thanks for the PR! I've marked one of my comments `IMPORTANT` and added a few minor nitpicks that would be good to address but aren't critical. Would you also be able to rebase off the `master` branch to resolve the conflicts in `config.go` and `go.sum`. Thanks!

galbirk commented:
Thanks @henrymcconville !
Will take a look and update. Hopefully in the next few days.

galbirk commented:
@henrymcconville I removed duplicated code and added some field to the Config struct.
I was thinking to create GitHubApp struct specifically for GitHubApp configuration to not overload the Config one. WDYT? For now, all added to the Config struct.

henrymcconville commented:
@galbirk Think a separate struct would be nicer but happy with it as is for now. Feel free to raise a follow up PR if you're passionate enough about it being separated 😄
